AMDGPU/GlobalISel: Regbanklegalize rules for G_UNMERGE_VALUES

petar-avramovic opened the pull request:
Move G_UNMERGE_VALUES handling to AMDGPURegBankLegalizeRules.cpp.
Fix sgpr S16 unmerge by lowering using shift and using S32.
Previously sgpr S16 unmerge was selected using _lo16 and _hi16 subreg
indexes which are exclusive to vgpr register classes.
For remaing cases we do trivial mapping, assigns same reg bank
to all operands, vgpr or sgpr.

petar-avramovic commented:
ping:
@broxigarchen, @kosarev, @Sisyph for true 16 part (subreg index for sgprs)
@arsenm for introduction of BRC, similar opcodes will also be rewritten using BRC, S16 and S1

mbrkusanin commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -910,6 +910,55 @@ bool RegBankLegalizeHelper::lower(MachineInstr &MI,
+      return false;
+    }
+
+    B.setInstr(MI);
```
```suggestion
    B.setInstrAndDebugLoc(MI);
```

mbrkusanin commented on llvm/lib/Target/AMDGPU/AMDGPUGlobalISelUtils.cpp:
```diff
@@ -186,3 +186,19 @@ void AMDGPU::buildReadFirstLane(MachineIRBuilder &B, Register SgprDst,
+  // 32, 2x32, 3x32 ... 12x32, 16x32, 32x32
+  unsigned NumB32s = Size / 32;
+  if ((NumB32s >= 1 && NumB32s <= 12) || NumB32s == 16 || NumB32s == 32)
+    return true;
```
I do not see a nice way to query list of supported sizes from tablegen or anywhere else. Maybe just add an assert after this check if something like `getSGPRClassForBitWidth` or `getRegClassForSizeOnBank` succeeds. That way if down the line a new size is introduced this will warn us to update values here as well. It can be guarded by `#ifndef NDEBUG`.

Unless you think returning false will end up producing clear enough error.

Sisyph reviewed (approved):
True16 changes LGTM

petar-avramovic committed:
AMDGPU/GlobalISel: Regbanklegalize rules for G_UNMERGE_VALUES

Move G_UNMERGE_VALUES handling to AMDGPURegBankLegalizeRules.cpp.
Fix sgpr S16 unmerge by lowering using shift and using S32.
Previously sgpr S16 unmerge was selected using _lo16 and _hi16 subreg
indexes which are exclusive to vgpr register classes.
For remaing cases we do trivial mapping, assigns same reg bank
to all operands, vgpr or sgpr.

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeRules.cpp:
```diff
@@ -176,6 +181,10 @@ bool matchUniformityAndLLT(Register Reg, UniformityLLTOpPredicateID UniID,
   case DivB512:
     return MRI.getType(Reg).getSizeInBits() == 512 && MUI.isDivergent(Reg);
+  case DivBRC:
+    return isVgprBRC(MRI.getType(Reg), static_cast<const SIRegisterInfo *>(
```
done
